### PR TITLE
fix(web-shell): don't steal approval focus while the user is typing

### LINKED ISSUE #9571: Avoid confirmation boxes being selected by default
### What happened?

During the execution of the webui task, if text is being entered in the input box, a confirmation box will pop up in the model, which will cause the confirmation box to be selected by default.

### What did you expect to happen?

During the execution of the webui task, if text is being entered in the input box, a confirmation box will pop up in the model, which will cause the confirmation box to be selected by default.

### Client information

<details>
<summary>Client Information</summary>

Run `qwen` to enter the interactive CLI, then run the `/about` command.

```console
$ qwen /about
Qwen Code v0.21.14
Model: deepseek-v4-flash-0731
Fast Model: not set
Auth: openai
Platform: win32 x64 (10.0.19045)
Node.js: v24.14.0
Session: 45bb998a-c557-48b3-81aa-071cf9199e6a
Git commit: 6e20a58923
LSP: disabled
```

</details>

start.bat:

```
@echo off
chcp 65001 >nul 2>&1
qwen serve --port 4170 --open --token='xxxxxxx' 
pause
```

### Login information

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by yiliang114**
## What this PR does

The Web Shell tool-approval dialog no longer pulls keyboard focus to its safe-default option when it appears while the user is typing in an editable element (the composer). The focus-grab effect in `ToolApproval` now yields when the active element is an editable target, reusing the existing `isEditableTarget` / `getShadowAwareActiveElement` helpers from `client/utils/dom.ts`. In every other situation — nothing focused, focus on a non-editable element, split-view panes, re-activation after a covering panel closes — the existing focus handoff is unchanged.

## Why it's needed

Fixes the behavior reported in #9571: while a task is running and the user is typing in the input box, the confirmation box pops up "selected by default". The mechanism: when a tool approval arrives, the same React commit hides the composer (`App.tsx`, `composerHidden`) and mounts the `ToolApproval` overlay, whose focus effect then unconditionally focuses the safe-default option button. Because keyboard handling is focus-scoped to the panel, the user's in-progress keystrokes land on the dialog: `Enter`/`Space` natively activate the focused default option, and digits `1`–`9` confirm by position — so the message-send Enter or plain typing can unintentionally approve or reject the tool call.

### How to verify

The fix rests on one browser-timing fact: when effects run (post-commit, pre-paint), the composer textarea still holds focus even though it was just given `display:none`; the browser drops that focus only during the later style recalc. Verified three ways:

1. **Red/green component tests** — `packages/web-shell/client/components/messages/ToolApproval.test.tsx`, new `yielding to active typing (#9571)` block (4 tests). On unfixed `main` all 4 fail (focus moves from the focused textarea to the safe-default option); with the fix all 36 tests in the file pass:

   ```text
   # before (unfixed main): Tests  4 failed | 32 passed (36)
   # after  (this branch):  Tests  36 passed (36)
   ```

2. **Real-browser timing probe** (Playwright headless Chromium): focus a textarea, then synchronously set its wrapper `display:none` and mount a button (the commit order), reading `document.activeElement` immediately and after paint:

   ```json
   { "focusedOnComposer": true, "syncActiveIsComposer": true, "syncActiveTag": "TEXTAREA", "postPaintActiveTag": "BODY" }
   ```

   i.e. at effect time the active element still reflects the typing target (guard works), and without any grab focus falls to `body` after paint — so stray keystrokes hit nothing instead of confirming the dialog.

3. **Regression scope** — full `packages/web-shell` suite: `Test Files 192 passed (192)`, `Tests 4001 passed (4001)` (includes the App-level approval-overlay and `composerHidden` tests, and the existing focus-grab/roving-tabindex/digit-shortcut tests which pin the unchanged behavior). `typecheck`, `eslint`, and `prettier --check` are clean.

A full manual pass (`qwen serve` + a live turn that triggers an approval while typing) was not run here — no model session is available in this environment; the component tests exercise the exact focus contract through real DOM focus (unmocked).

### Evidence (Before & After)

Focus behavior is not screenshot-capturable; the red/green test block and the browser probe output above are the evidence. Before: `document.activeElement` jumps from the focused composer textarea to the `Reject`/safe-default option the instant the approval mounts. After: it stays on the textarea, and the dialog is reached by Tab/click as designed.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  |   ⚠️   |
| 🪟 Windows |   ⚠️   |
|  🐧 Linux  |   ✅   |

### Environment (optional)

Unit tests: `npx vitest run --config vitest.config.ts` in `packages/web-shell` (Node v24.19.0). Timing probe: Playwright headless Chromium.

## Risk & Scope

- Main risk or tradeoff: when an approval appears exactly while the user is typing, keyboard users are not auto-placed on the default option and need one Tab into the dialog; appearance without active typing is unchanged. If a browser ever dropped focus from the hidden composer before effects run, the guard simply misses and behavior degrades to today's grab — never worse than status quo.
- Not validated / out of scope: `AskUserQuestion` has the symmetric unguarded focus grab (its key *shortcuts* already check `isEditableTarget`, the grab does not) — left for a follow-up to keep this PR scoped to the reported confirmation box. No live e2e with a real daemon/model session.
- Breaking changes / migration notes: none.

## Linked Issues

Fixes #9571

<details>
<summary>中文说明</summary>

## 这个 PR 做了什么

当工具审批弹窗出现时，如果用户正在可编辑元素（输入框）中打字，Web Shell 的审批弹窗不再把键盘焦点抢到"安全默认"选项上。`ToolApproval` 的焦点抢占 effect 现在会在当前焦点位于可编辑元素时让位，复用 `client/utils/dom.ts` 里现成的 `isEditableTarget` / `getShadowAwareActiveElement` 工具函数。其余所有情况——无焦点、焦点在非编辑元素、分栏面板、遮挡面板关闭后的重新激活——原有的焦点交接行为完全不变。

## 为什么需要

修复 #9571 报告的行为：任务执行中用户在输入框打字时，确认弹窗弹出即"已默认选中"。机制是：工具审批到达时，同一个 React commit 里既隐藏了输入框（`App.tsx` 的 `composerHidden`）又挂载了 `ToolApproval` 弹窗，其焦点 effect 随后无条件地把焦点移到安全默认选项按钮上。由于键盘处理是按面板焦点作用域绑定的，用户正在敲的按键就落在弹窗上：`Enter`/`Space` 原生激活被聚焦的默认选项，数字键 `1`–`9` 按位置直接确认——于是习惯性的"回车发送"或正在输入的字符可能无意间批准或拒绝工具调用。

### 如何验证

修复依赖一个浏览器时序事实：effect 运行时（commit 之后、绘制之前），即使输入框刚被设为 `display:none`，焦点仍在输入框 textarea 上；浏览器要到之后的样式重算才会丢掉这个焦点。用三种方式验证：

1. **红转绿组件测试**——`packages/web-shell/client/components/messages/ToolApproval.test.tsx` 新增 `yielding to active typing (#9571)` 块（4 条）。未修复的 `main` 上 4 条全部失败（焦点从已聚焦的 textarea 跳到安全默认选项）；修复后该文件 36 条全部通过。

2. **真实浏览器时序探针**（Playwright headless Chromium）：聚焦一个 textarea，然后同步地把它的外层设为 `display:none` 并挂载一个按钮（即 commit 顺序），分别在当时和绘制后读取 `document.activeElement`：effect 时刻焦点仍是 textarea（守卫有效），绘制后焦点落到 `body`——即不抢占时多余按键什么都不会命中，而不是确认弹窗。

3. **回归范围**——`packages/web-shell` 全量套件：192 个文件 / 4001 条测试全部通过（含 App 层审批弹窗与 `composerHidden` 测试，以及固化原有行为的焦点抢占/roving tabindex/数字快捷键测试）。`typecheck`、`eslint`、`prettier --check` 均干净。

没有跑完整的手动链路（`qwen serve` + 真实触发审批的回合 + 打字）——本环境没有可用模型会话；组件测试通过真实 DOM 焦点（未 mock）覆盖的正是这条焦点契约。

### 前后证据

焦点行为无法截图取证；上面的红转绿测试块与浏览器探针输出即证据。修复前：审批弹窗挂载瞬间 `document.activeElement` 从已聚焦的输入框跳到安全默认选项。修复后：焦点保持在输入框，弹窗按设计通过 Tab/点击到达。

### 测试环境

Linux ✅（单测 + 探针）；macOS / Windows 未测。

## 风险与范围

- 主要风险/取舍：审批恰好在用户打字时出现时，键盘用户不会被自动放到默认选项上，需要多按一次 Tab 进入弹窗；无输入行为时的弹窗出现不受影响。若某浏览器在 effect 运行前就丢掉了被隐藏输入框的焦点，守卫只会失效并退回今天的抢占行为——不会比现状更差。
- 未验证/超出范围：`AskUserQuestion` 存在同样未加守卫的焦点抢占（它的按键快捷键已有 `isEditableTarget` 检查，抢占本身没有）——为把本 PR 收敛在报告所述的确认框，留作后续处理。未做真实 daemon/模型会话的 live e2e。
- 破坏性变更/迁移说明：无。

</details>

### CODE REVIEW

File: packages/web-shell/client/components/messages/ToolApproval.tsx [hunk 1]
```diff
@@ -316,6 +318,18 @@ export function ToolApproval({
...
+    // still on the editable target when this effect runs. Grabbing it would
+    // redirect the in-progress keystrokes — Enter-to-send, Space, digits — to
+    // the safe-default option and can confirm the request unintentionally.
+    // Yield to the editable target; the dialog stays reachable by Tab/click.
+    // (The composer is hidden via CSS in the same commit, so style recalc —
+    // and its focus drop — has not happened yet when effects run; the active
+    // element still reflects where the user was typing.)
+    if (isEditableTarget(getShadowAwareActiveElement(panelRef.current))) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (yiliang114) [reply packages/web-shell/client/components/messages/ToolApproval.tsx:337]: Fixed in 155efb6bc0 — adopted the probe-validated option: the focus-grab effect is now a `useLayoutEffect`. Layout effects run synchronously after DOM mutation, before any forced style recalculation (e.g. `MessageList`'s messages-keyed `scrollToBottom` reading `scrollHeight`), so the guard always evaluates while the just-hidden composer still holds focus — your arm C shape. Comment above the effect now documents why a passive effect is unsound here. jsdom cannot distinguish the two timings (it never blurs on `display:none`), so the red/green evidence is your 5/5 Chromium probe arm C plus the existing 4 yield tests staying green (38/38 in the file); the effect body sets no state, so the layout-effect timing adds no synchronous re-render path.

[the remark this answers is not in the document]
Comment (yiliang114) [reply packages/web-shell/client/components/messages/ToolApproval.tsx:337]: Deferred by design — the PR description explicitly scopes this PR to the reported confirmation box ("left for a follow-up to keep this PR scoped"). Filed the follow-up as #9611 with the exact guard shape suggested here (panel-root ref + `isEditableTarget(getShadowAwareActiveElement(...))` early-out, run as a layout effect per the timing finding above). Closing this thread with the issue tracked.

File: packages/web-shell/client/components/messages/ToolApproval.test.tsx [hunk 2]
```diff
@@ -266,6 +266,61 @@ describe('ToolApproval accessibility', () => {
...
+    let composer: HTMLTextAreaElement;
+
+    beforeEach(() => {
+      // The user is typing in the composer when the approval arrives. In the
+      // app the overlay commit hides the composer in the same render, so the
+      // editable target still holds focus when the focus effect runs (jsdom
+      // mirrors that: display:none never blurs).
+      composer = document.createElement('textarea');
+      document.body.appendChild(composer);
+      composer.focus();
```
[the remark this answers is not in the document]
Comment (yiliang114) [reply packages/web-shell/client/components/messages/ToolApproval.test.tsx:279]: Added in 155efb6bc0 — new yield case `yields to a contenteditable composer (CodeMirror shape)` focuses a contenteditable div inside a `.cm-editor` wrapper (the production EditorView shape) and asserts no option receives focus. Mutation-verified: stripping the contenteditable branches from `isEditableTarget` (the `isContentEditable` check plus the `[contenteditable="true"], .cm-editor` selector entries) fails exactly this test; restored code passes (38/38).

File: packages/web-shell/client/components/messages/ToolApproval.test.tsx [hunk 3]
```diff
@@ -266,6 +266,61 @@ describe('ToolApproval accessibility', () => {
     );
   });
 
+  describe('yielding to active typing (#9571)', () => {
```
[the remark this answers is not in the document]
Comment (yiliang114) [reply packages/web-shell/client/components/messages/ToolApproval.test.tsx:269]: Added in 155efb6bc0 — new yield case `yields in shadow-DOM (portal) mode` mounts `ToolApproval` via `createRoot` into a div inside `host.attachShadow({mode:'open'})`, focuses a textarea in the same shadow root, and asserts `shadowRoot.activeElement` stays the textarea with no option focused (modeled on the `AddWorkspaceDialog.test.tsx` attachShadow precedent). Mutation-verified: replacing `getShadowAwareActiveElement(panelRef.current)` with `document.activeElement` fails exactly this test (the host-retarget shape you described); restored code passes (38/38).

**Comment by yiliang114** [comment]
Closeout of the 21:53Z round on 458398e550 — all 4 findings handled, pushed 155efb6bc0 (one non-force push).

**Critical FIXED** — the yield guard now runs in a `useLayoutEffect` (`ToolApproval.tsx`). Adopted the finding's probe-validated option (arm C): layout effects execute synchronously after DOM mutation and before any forced style recalculation, so the guard evaluates while the just-hidden composer still holds focus, regardless of sibling layout-reading effects (`MessageList` `scrollToBottom`). A comment above the effect records why a passive effect is unsound here. No manual Before/After screenshots: focus-timing races are not screenshot-capturable (as the PR body notes) — the evidence is the finding's own deterministic 5/5 Chromium probe of this exact change plus the yield tests below.

**Suggestion FIXED** — new test `yields to a contenteditable composer (CodeMirror shape)`: focuses a contenteditable div inside `.cm-editor` (production EditorView shape) and pins `isEditableTarget`'s contenteditable branches. Mutation-verified (stripping those branches fails exactly this test).

**Suggestion FIXED** — new test `yields in shadow-DOM (portal) mode`: mounts into `attachShadow({mode:'open'})` per the `AddWorkspaceDialog` precedent and pins the `getShadowAwareActiveElement` branch. Mutation-verified (swapping in `document.activeElement` fails exactly this test).

**Suggestion deferred by design** — the `AskUserQuestion` mirror guard: the PR description explicitly scopes this PR to the reported confirmation box; filed #9611 to track it with the suggested guard shape.

Verification: `ToolApproval.test.tsx` 38/38, web-shell `tsc --noEmit` clean, eslint + prettier clean on both touched files. No UI-surface change beyond the guard's timing (behavior when the guard misses is unchanged).

**Comment by yiliang114** [comment]
@qwen-code /triage

**Comment by yiliang114** [comment]
@qwen-code /review --timeout=360

**Review by doudouOUC — COMMENTED** [review]
## Review: `fix(web-shell): don't steal approval focus while the user is typing`

**Overall**: Clean, focused fix. The core approach — yielding the focus effect when the active element is an editable target — is correct and well-implemented. The `useLayoutEffect` change is justified by the Chromium timing behavior. Test coverage is thorough. No criticals.

---

### Correctness

- **The `useLayoutEffect` rationale is sound.** The comment explains why a passive `useEffect` would miss the focus because Chrome drops focus from the hidden composer during pre-paint style recalculation, before passive effects run. Layout effects run right after DOM mutation, before that recalculation, so the guard still sees the composer as focused. Verified by reading the code.

- **Shadow DOM handling is correct.** `getShadowAwareActiveElement(panelRef.current)` resolves the active element from the panel's own root, correctly handling the `document.activeElement` retargeting that happens when focus is inside a shadow root. This is the right approach.

- **`panelRef` is set before the layout effect runs.** React assigns refs during the commit phase, before layout effects fire. The edge case where `panelRef.current` is null is a safe fallback (the old behavior, no-op against the guard).

### Test Quality

- **6 test cases covering the main scenarios**: textarea focus, re-render mid-typing, re-activation, tab-in interaction, contenteditable (CodeMirror shape), and shadow DOM. Good coverage.

- **The contenteditable test is a nice touch**: it pins the CodeMirror shape specifically, which is the actual production composer — not just the textarea backend.

- **The shadow DOM test is valuable**: portal mode is a real deployment scenario, and testing the retargeting behavior is important.

### Suggestions (non-blocking)

1. **The `isEditableTarget` helper in `dom.ts` includes `[data-keyboard-scope]`** — this selector matches elements that are not inherently editable. The name `isEditableTarget` is slightly misleading for that case. However, this is pre-existing code, not introduced by this PR, so it's not actionable here.

2. **The `useLayoutEffect` comment (lines 314–321 in the diff) is thorough but could be slightly more concise.** The technical detail about Chromium's style-recalculation timing is important, but the first sentence ("Must be a layout effect, not a passive one") already tells the reader the decision. The rest repeats the same observation. Not a change request — just observing that the comment could be tightened without losing information.

3. **The CI bot's suggestion** about a timing-discriminating test (rendering a `BlurComposer` sibling that blurs the composer in a passive effect) is valid but not a blocker — jsdom doesn't replicate the Chromium timing difference, so such a test would only be meaningful in a real browser environment. The existing tests are sufficient for the jsdom level.

### Summary

**No critical or blocking issues.** The PR is well-factored, the fix is minimal and precise, and the tests are thorough. Ready to ship.

_— Manually reviewed_

**Review by chiga0 — APPROVED** [review]
No blocking findings.

**Checked:**
- Timing correctness: `useEffect` → `useLayoutEffect` is the right fix. At layout-effect time the commit that hides the composer and mounts the overlay hasn't triggered a style recalculation yet — the browser still reports the textarea as `document.activeElement`. A passive effect would run after the recalculation and would see `body` instead, making the guard a no-op. The Playwright timing probe in the PR confirms the synchronous-recalc path.
- Guard implementation: `isEditableTarget(getShadowAwareActiveElement(panelRef.current))` — `panelRef.current` is set during the commit phase, before layout effects run, so it is never null at guard time. Shadow DOM path covered by test.
- Coverage: textarea, `contenteditable` (CodeMirror shape), shadow-DOM portal, re-activation, new-request-mid-typing — all pinned by new tests. Existing roving-tabindex / digit-shortcut / focus-grab tests pass unchanged.
- Fallback behavior: if the browser ever drops focus before the layout effect fires, the guard misses and behavior reverts to the old grab — no regression path.

**CI:** ubuntu Test + E2E Smoke + visual capture → all green. macOS/Windows SKIPPED (repo pattern).

**Not covered:** macOS/Windows matrix; `AskUserQuestion` symmetric gap (acknowledged in PR body, left for follow-up).


_Reviewed with AI assistance._
